### PR TITLE
HTML Search: omit anchor reference from document titles in the search index.

### LINKED ISSUE #11961: HTML Search: Contains duplicates based on title and content search
### Describe the bug

We currently list the same page multiple times if, for example, both the title and content search match it. Since the preview is always the same, this is not very helpful.

### How to Reproduce

An easy way to reproduce the bug is to do a title search: the first result will link to the section, the second will link to the page in general. It appears that the "remove duplicates" feature [was added some time ago](https://github.com/sphinx-doc/sphinx/commit/e3a8744235adeefdecc1160c0fadfa2322c4bf7d), perhaps before there were multiple ways in which documents were searched.

https://pradyunsg.me/furo/search/?q=quickstart&check_keywords=yes&area=default

Screenshot from 2024-02-04 13-42-21

Note that it appears the main sphinx documentation uses readthedocs' custom search, so this bug wouldn't be reproducible there.

### Environment Information

Sphinx main as of 2024-02-06: https://github.com/sphinx-doc/sphinx/commit/e976059fd6ab49b1c2784445d7095c9287897724

### Sphinx extensions

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jayaddison**
### Feature or Bugfix
- Bugfix

### Purpose
- Adjusts the contents of the search-index provided to the client so that the browser correctly de-duplicates search results that link to the top-level title in documents.

### Detail
- Adapts the transformation of `docutils` `nodes.title` nodes into the title index by omitting their `id` that is used as the hyperlink anchor in the special-case of the document title.
- Note: there is [no explicit markup for the document title in reStructuredText](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#document).

### Relates
- Resolves #11961.
- Alternative approach to client-side fix in #11942.

### CODE REVIEW

**Comment by wlach** [comment]
This seems preferable to my solution in #11942, just tested and it has the same results while making the search index smaller. :+1: from me.

**Comment by picnixz** [comment]
Fine for me (I'm not an expert in that search-related aspect so I'll leave it to you guys). Should I understand that this PR would replace #11942 ? @jayaddison ping me when you want it to be merged

**Comment by jayaddison** [comment]
> Fine for me (I'm not an expert in that search-related aspect so I'll leave it to you guys). Should I understand that this PR would replace #11942 ?

That's correct.  I think there are three things that I don't like about this pull request, in order of priority:

1. It doesn't offer JavaScript test coverage to demonstrate the fix -- only Python test coverage that confirms the expected index format.
2. I haven't been able to think of a way that this would happen yet, but in theory I expect some downstream tools that use the current index format could be broken by this - in particular the possibility of `null` values where previously everything was a string.
3. Related to (2), the index format isn't as small as it could be.  In raw text, the JavaScript empty string can be represented by either `''` or `""` (two characters), whereas a null value is represented by the token `null` (four characters).  It may not matter a lot, but over a long enough duration of time the cumulative cost could be significant (or not!  maybe it's a waste of time considering it).

**Comment by jayaddison** [comment]
> It doesn't offer JavaScript test coverage to demonstrate the fix -- only Python test coverage that confirms the expected index format.

I think I'll begin work on a small JavaScript refactor PR to make test coverage easier to add.

**Comment by picnixz** [comment]
> particular the possibility of null values where previously everything was a string.

Is it possible to keep a string or is the null type needed?

**Comment by jayaddison** [comment]
> > particular the possibility of null values where previously everything was a string.
> 
> Is it possible to keep a string or is the null type needed?

If I remember correctly, [this conditional](https://github.com/sphinx-doc/sphinx/blob/5695b76f05af253f8eb3a956d25950af81befc7e/sphinx/themes/basic/static/searchtools.js#L310) needs to be adjusted if we're using empty strings, but otherwise it should be fine.

What I would really like is a way to generate the indexes used in the JavaScript tests from the same Python code that builds `searchindex.js` when projects are built.  It feels fragile that the test indexes are written by hand.

**Comment by jayaddison** [comment]
> What I would really like is a way to generate the indexes used in the JavaScript tests from the same Python code that builds `searchindex.js` when projects are built. It feels fragile that the test indexes are written by hand.

Moved into #12099.

**Comment by jayaddison** [comment]
I'd like to check whether we can get #12102 in place before progressing this pull request further.  If that can be added, then I think adding test coverage here will be much easier and more reliable (I'll be able to create a sample Sphinx project that returns duplicate search results, and add test coverage against that).

**Comment by jayaddison** [comment]
(and maybe do the `null` to `''` refactoring at the same time and more safely, given the test coverage)

**Comment by picnixz** [comment]
Removing the "awaiting review" label until this PR is ready

**Comment by jayaddison** [comment]
> Removing the "awaiting review" label until this PR is ready

Oops, thanks.  I forgot about that.

**Comment by jayaddison** [comment]
While I think it would be sensible to merge #12102 first (if-and-when that pull request is considered acceptable!), I'm going to remove the blocker label and draft status from this pull request.

**Comment by jayaddison** [comment]
@picnixz I think this one is now ready for merge too - glad to have #12102 already-merged as preparation.

File: sphinx/search/__init__.py
Comment (picnixz) [inline sphinx/search/__init__.py]: Just split the line into to (in general, I only use E501 for signatures, not for calls like that).

**Comment by jayaddison** [comment]
Hmm, the type-linting error highlighted by that adjustment is not entirely trivial; it bubbles up to the `BuildEnvironment` class.  I'll fix it, but I seem to remember that environment changes can affect pickling (and thus project build caches?).

**Comment by jayaddison** [comment]
(similarly it could be worth pausing to consider `searchindex.js` forward/backward compatibility.  I _think_ this change is safe because there are no associated JavaScript changes required for this fix, but that may be naive)

**Comment by wlach** [comment]
> Hmm, the type-linting error highlighted by that adjustment is not entirely trivial; it bubbles up to the `BuildEnvironment` class. I'll fix it, but I seem to remember that environment changes can affect pickling (and thus projechttps://docs.github.com/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/about-protected-branchest build caches?).

Hmm, my understanding is that under normal circumstances type changes do not affect runtime behaviour at all.

> (similarly it could be worth pausing to consider searchindex.js forward/backward compatibility. I think this change is safe because there are no associated JavaScript changes required for this fix, but that may be naive)

IMO it's fine. We're just changing what gets included in the searchindex.

**Comment by jayaddison** [comment]
> > Hmm, the type-linting error highlighted by that adjustment is not entirely trivial; it bubbles up to the `BuildEnvironment` class. I'll fix it, but I seem to remember that environment changes can affect pickling (and thus projechttps://docs.github.com/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/about-protected-branchest build caches?).
> 
> Hmm, my understanding is that under normal circumstances type changes do not affect runtime behaviour at all.

Mm, yep - the type hints themselves shouldn't affect anything... but they are an indicator that something about the possible range of values for their attributes has changed (`None` now permitted) - and since the [`env` is serialized](https://github.com/sphinx-doc/sphinx/blob/de4ac2fbdebae529a973910e96794645158f0401/sphinx/builders/__init__.py#L327-L329) and potentially re-loaded during non-fresh builds (to speed up project rebuilds), I'm worrying about possible regressions due to that.

> > (similarly it could be worth pausing to consider searchindex.js forward/backward compatibility. I think this change is safe because there are no associated JavaScript changes required for this fix, but that may be naive)
> 
> IMO it's fine. We're just changing what gets included in the searchindex.

I think I agree here - it's simply that document main titles are now linked directly instead of with a hyperlink anchor/target.

However, similar to the `env` pickling/reloading -- [there is code](https://github.com/sphinx-doc/sphinx/blob/de4ac2fbdebae529a973910e96794645158f0401/sphinx/search/__init__.py#L295-L299) that I remember seeing that can reload the search state from a previously-frozen serialized format.

Basically, in both cases there is some element of possible stateful data that downstream projects may have written to disk, and so I want to be careful to make sure that the changes we introduce are at-least forwards compatible (deploying an updated version of Sphinx and rebuilding a project does not fail), and perhaps even backwards compatible where possible (rolling back to a previous version of Sphinx after a rebuild with a newer version would also succeed).

File: tests/test_search.py
Comment (picnixz) [inline tests/test_search.py]:
```suggestion
another_title
=============
```

File: tests/test_search.py
Comment (picnixz) [inline tests/test_search.py:186]: Here, I see that the values are sets and thus can be unordered. Is it possible to actually have a more reliable guarantee on the `title_mapping` actually?

File: tests/test_search.py
Comment (picnixz) [inline tests/test_search.py:187]: Can you use a shorter search term? without the "e", I almost thought it was a typo and wanted to correct it. For instance, how about `another_ti`? (same question for `section_*` and other cases)

File: sphinx/search/__init__.py
Comment (picnixz) [inline sphinx/search/__init__.py:401]: Could you explain the shape of this value, i.e., to what corresponds `list[tuple[int, str | None]` (what is the integer, what is the `str | None` and) and what is the key `str` of the `dict`, and why does it change from the `all_titles` ? (the list has pairs of int and strings and not pairs of strings anymore).

File: tests/test_search.py
Comment (jayaddison) [reply tests/test_search.py:187]: In this case, `another_titl` isn't a user-entered query, but is the [stemmed](https://en.wikipedia.org/wiki/Stemming) form of the title that would be stored in the index file (`searchindex.js`).

We could try to modify the test documents so that they use heading names that look like valid English even after stemming - what do you think?

File: tests/test_search.py
Comment (jayaddison) [reply tests/test_search.py:186]: To check that I understand correctly: use an assertion that checks not only the _contents_ of the title mapping, but also the _ordering_ of the elements inside it?

(makes sense to me, order can be significant for index data)

File: sphinx/search/__init__.py
Comment (jayaddison) [reply sphinx/search/__init__.py:401]: Mm, this is not very clear at the moment, but I believe that the integer values here are what I'd call `documentId` values - sequential IDs assigned to each of the source `.rst` files at build-time.  The second value in each tuple -- the `str` or empty value -- is a target reference (aka `anchor` when talking about HTML) -- it's an indicator about what destination on the page/document the user should be taken to were they to navigate using this entry.

Perhaps this is a situation where we should refactor and extract out meaningful type aliases (a bit like the recent Intersphinx refactorings).

File: tests/test_search.py
Comment (picnixz) [reply tests/test_search.py:187]: Oh, no, let's not overcomplicate the things. I think a document telling us (people who don't work on that component) how the test API works would be good actually, like a README, or something like that, but it should be another PR.

File: sphinx/search/__init__.py
Comment (picnixz) [reply sphinx/search/__init__.py:401]: I would **not** use aliases, but the `int` always seem to me like corresponding to the index in the list, so that you could possibly have multiple times the same element but with different indices. However, I'd be happy to know if this is really the reason why we duplicate that or if we could just live with a list of strings.

Comment (jayaddison) [reply sphinx/search/__init__.py:401]: I guess an example to exlain that could be a title like `Recent Changes` that could appear in multiple documents and might have different headings (or perhaps even be the document title in some cases).

`alltitles: {"Recent Changes": [0, None], [1, "coolmodule-recentchanges"], [5, "anothermodule-recentchanges"], ...}`

(a bit hypothetical, but it's to handle situations like that I believe.  again, a good case for checking and documenting)

**Comment by wlach** [comment]
> Mm, yep - the type hints themselves shouldn't affect anything... but they are an indicator that something about the possible range of values for their attributes has changed (`None` now permitted) - and since the [`env` is serialized](https://github.com/sphinx-doc/sphinx/blob/de4ac2fbdebae529a973910e96794645158f0401/sphinx/builders/__init__.py#L327-L329) and potentially re-loaded during non-fresh builds (to speed up project rebuilds), I'm worrying about possible regressions due to that.

I'm struggling a bit to see the problem: aside from deserialization, I can't imagine for what reason we'd be reading index entries from Python code? I skimmed through the code and don't see us doing anything where the different types would be a problem:

https://github.com/search?q=repo%3Asphinx-doc%2Fsphinx%20self._titles&type=code

> > > (similarly it could be worth pausing to consider searchindex.js forward/backward compatibility. I think this change is safe because there are no associated JavaScript changes required for this fix, but that may be naive)
> 
> > 
> > IMO it's fine. We're just changing what gets included in the searchindex.
> 
> I think I agree here - it's simply that document main titles are now linked directly instead of with a hyperlink anchor/target.
> 
> However, similar to the `env` pickling/reloading -- [there is code](https://github.com/sphinx-doc/sphinx/blob/de4ac2fbdebae529a973910e96794645158f0401/sphinx/search/__init__.py#L295-L299) that I remember seeing that can reload the search state from a previously-frozen serialized format.

It looks like we explicitly reject any index created with a previous version of Sphinx in this case:

https://github.com/sphinx-doc/sphinx/blob/de4ac2fbdebae529a973910e96794645158f0401/sphinx/search/__init__.py#L302

**Comment by jayaddison** [comment]
> > Mm, yep - the type hints themselves shouldn't affect anything... but they are an indicator that something about the possible range of values for their attributes has changed (`None` now permitted) - and since the [`env` is serialized](https://github.com/sphinx-doc/sphinx/blob/de4ac2fbdebae529a973910e96794645158f0401/sphinx/builders/__init__.py#L327-L329) and potentially re-loaded during non-fresh builds (to speed up project rebuilds), I'm worrying about possible regressions due to that.
> 
> I'm struggling a bit to see the problem: aside from deserialization, I can't imagine for what reason we'd be reading index entries from Python code? I skimmed through the code and don't see us doing anything where the different types would be a problem:
> 
> https://github.com/search?q=repo%3Asphinx-doc%2Fsphinx%20self._titles&type=code

I think the theory is that the search index -- like individual output files in many cases -- could be rebuilt incrementally rather than having to re-read the entire project content.

Ref: b1271fa623f77140d8a96e7e29514a4f75cf1a64

However, it's not clear to me whether those loaded entries are retained before the updated search index is written, because of the code linked below; is `env.all_docs` the _entire project_ (in which case it's definitely not an incremental rebuild), or are they only the modified documents?

https://github.com/sphinx-doc/sphinx/blob/de4ac2fbdebae529a973910e96794645158f0401/sphinx/builders/html/__init__.py#L1163

> > > > (similarly it could be worth pausing to consider searchindex.js forward/backward compatibility. I think this change is safe because there are no associated JavaScript changes required for this fix, but that may be naive)
> > 
> > 
> > > IMO it's fine. We're just changing what gets included in the searchindex.
> > 
> > 
> > I think I agree here - it's simply that document main titles are now linked directly instead of with a hyperlink anchor/target.
> > However, similar to the `env` pickling/reloading -- [there is code](https://github.com/sphinx-doc/sphinx/blob/de4ac2fbdebae529a973910e96794645158f0401/sphinx/search/__init__.py#L295-L299) that I remember seeing that can reload the search state from a previously-frozen serialized format.
> 
> It looks like we explicitly reject any index created with a previous version of Sphinx in this case:
> 
> https://github.com/sphinx-doc/sphinx/blob/de4ac2fbdebae529a973910e96794645158f0401/sphinx/search/__init__.py#L302

Ok, that's good to know - but that version number is a static value in the codebase -- it's not the Sphinx version number, nor a counter of the number of builds on the machine where Sphinx is being run:

https://github.com/sphinx-doc/sphinx/blob/de4ac2fbdebae529a973910e96794645158f0401/sphinx/environment/__init__.py#L59-L61

**Comment by wlach** [comment]
> > I'm struggling a bit to see the problem: aside from deserialization, I can't imagine for what reason we'd be reading index entries from Python code? I skimmed through the code and don't see us doing anything where the different types would be a problem:
> > https://github.com/search?q=repo%3Asphinx-doc%2Fsphinx%20self._titles&type=code
> 
> I think the theory is that the search index -- like individual output files in many cases -- could be rebuilt incrementally rather than having to re-read the entire project content.
> 
> Ref: [b1271fa](https://github.com/sphinx-doc/sphinx/commit/b1271fa623f77140d8a96e7e29514a4f75cf1a64)

Ok, but I don't see anything in the codebase accessing that variable (`self._titles`) aside from when we write per that search query above. Let me know if I'm missing something.

> > It looks like we explicitly reject any index created with a previous version of Sphinx in this case:
> > https://github.com/sphinx-doc/sphinx/blob/de4ac2fbdebae529a973910e96794645158f0401/sphinx/search/__init__.py#L302
> 
> Ok, that's good to know - but that version number is a static value in the codebase -- it's not the Sphinx version number, nor a counter of the number of builds on the machine where Sphinx is being run:
> 
> https://github.com/sphinx-doc/sphinx/blob/de4ac2fbdebae529a973910e96794645158f0401/sphinx/environment/__init__.py#L59-L61

You could always increment that number if you wanted to be on the safe side. Not really a lot of downside outside of maybe a very minor build-time penalty in a few obscure cases.

**Comment by jayaddison** [comment]
> > > I'm struggling a bit to see the problem: aside from deserialization, I can't imagine for what reason we'd be reading index entries from Python code? I skimmed through the code and don't see us doing anything where the different types would be a problem:
> > > https://github.com/search?q=repo%3Asphinx-doc%2Fsphinx%20self._titles&type=code
> > 
> > 
> > I think the theory is that the search index -- like individual output files in many cases -- could be rebuilt incrementally rather than having to re-read the entire project content.
> > Ref: [b1271fa](https://github.com/sphinx-doc/sphinx/commit/b1271fa623f77140d8a96e7e29514a4f75cf1a64)
> 
> Ok, but I don't see anything in the codebase accessing that variable (`self._titles`) aside from when we write per that search query above. Let me know if I'm missing something.

Ok - that matches my understanding too; the only time we write to the search-related attributes on `BuildEnvironment` is during loading of an existing `searchindex.js` from the output build directory.  Subsequently some/all of that in-memory content is invalidated before we write the updated `searchindex.js` to that same output build directory.

> > > It looks like we explicitly reject any index created with a previous version of Sphinx in this case:
> > > https://github.com/sphinx-doc/sphinx/blob/de4ac2fbdebae529a973910e96794645158f0401/sphinx/search/__init__.py#L302
> > 
> > 
> > Ok, that's good to know - but that version number is a static value in the codebase -- it's not the Sphinx version number, nor a counter of the number of builds on the machine where Sphinx is being run:
> > https://github.com/sphinx-doc/sphinx/blob/de4ac2fbdebae529a973910e96794645158f0401/sphinx/environment/__init__.py#L59-L61
> 
> You could always increment that number if you wanted to be on the safe side. Not really a lot of downside outside of maybe a very minor build-time penalty in a few obscure cases.

Given that we believe that the Python-side changes should be backwards-compatible here (no structural changes to the pickled `env` data stored-and-read by Sphinx), I think I'd prefer to leave the env-version value as-is.

**Comment by jayaddison** [comment]
NodeJS tests on this branch began failing from 9fef43ba3ed3c5cb09a264994b55dc1d124e7810 onwards; that isn't easily visible from the CI results in the GitHub web UI because we weren't running those on each commit until #12445 was merged (the latest merge-from-master includes that, explaining the visibility of the test failure).

The change at 9fef43ba3ed3c5cb09a264994b55dc1d124e7810 regenerated the search index file for the repro test case.  Next I'm going to try to figure out why the tests began failing after that, and what that means.

**Comment by jayaddison** [comment]
Good news, I think.

> NodeJS tests on this branch began failing from [9fef43b](https://github.com/sphinx-doc/sphinx/commit/9fef43ba3ed3c5cb09a264994b55dc1d124e7810) onwards; that isn't easily visible from the CI results in the GitHub web UI because we weren't running those on each commit until #12445 was merged (the latest merge-from-master includes that, explaining the visibility of the test failure).
> 
> The change at [9fef43b](https://github.com/sphinx-doc/sphinx/commit/9fef43ba3ed3c5cb09a264994b55dc1d124e7810) regenerated the search index file for the repro test case. Next I'm going to try to figure out why the tests began failing after that, and what that means.
> NodeJS tests on this branch began failing from [9fef43b](https://github.com/sphinx-doc/sphinx/commit/9fef43ba3ed3c5cb09a264994b55dc1d124e7810) onwards; that isn't easily visible from the CI results in the GitHub web UI because we weren't running those on each commit until #12445 was merged (the latest merge-from-master includes that, explaining the visibility of the test failure). 
> 
> The change at [9fef43b](https://github.com/sphinx-doc/sphinx/commit/9fef43ba3ed3c5cb09a264994b55dc1d124e7810) regenerated the search index file for the repro test case. Next I'm going to try to figure out why the tests began failing after that, and what that means.

This index-regeneration occurred directly after e947f6b2e4b1023b45679816167fbed28067177c - a merge from the mainline branch.  That merge pulled in support for JavaScript testing of client-side HTML.  And _that_ happened to include a test case -- with a [relevant TODO note](https://github.com/sphinx-doc/sphinx/blob/e947f6b2e4b1023b45679816167fbed28067177c/tests/js/searchtools.js#L73) -- indicating that some of the data in it is a workaround and should be removed when #11961 is fixed.  _This PR_ fixes #11961 - so it is in fact expected that we should remove that data, and it's good that the test is failing while it is still in place.

File: tests/test_search.py
Comment (jayaddison) [reply tests/test_search.py:186]: If my understanding there was correct: I'd prefer to handle that in a separate issue thread and PR, if that's OK.  Or, if I misunderstood: let's make sure to resolve that first.

Comment (picnixz) [reply tests/test_search.py:186]: If there is more than one occurrence of that, yes we could discuss it in another thread.

Comment (jayaddison) [reply tests/test_search.py:186]: Yep; I think it would be applicable for both `_title_mapping` and `_mapping`; because both of those currently use a pattern of `<mapping>.setdefault(..., set()).add(...)` (implying that insertion-order may not match subsequent access iteration order):

https://github.com/sphinx-doc/sphinx/blob/b70578f80c78b00c6571203f3204126742367bce/sphinx/search/__init__.py#L447-L464

Comment (jayaddison) [reply tests/test_search.py:186]: Correction: any _runtime_ (in-memory) differences in the ordering of objects within the `_title_mapping` and `_mapping` set entries are later resolved into deterministic order by an [explicit sort](https://github.com/sphinx-doc/sphinx/blob/b70578f80c78b00c6571203f3204126742367bce/sphinx/search/__init__.py#L380) that occurs during [freezing](https://github.com/sphinx-doc/sphinx/blob/b70578f80c78b00c6571203f3204126742367bce/sphinx/search/__init__.py#L388) (serialization) of the search index (thanks to d24bd73d0ce8071459a1056691a1934e1ca12194).

I'm not aware of any other current bugs causing nondeterminism in our `searchindex.js` output since v7.3.0 (95fb0e5e570481cfb5e98990f657e3ef0fc26adc included in that release handled a similar situation).

**Comment by wlach** [comment]
@jayaddison Is there anything left to be done here or is this ready for a review by a committer?

**Comment by jayaddison** [comment]
This is ready, yep; the `mypy` lint failure on the latest commit (64387c4ae9a10cbd6851a4a693b21b00d369511b) is unrelated (#12510), and my last two messages where I'm talking to myself (not uncommon, sadly) are because I hadn't updated the test coverage after merging from mainline (to remove a duplicate result expectation added for a separate, already-merged PR).

**Comment by picnixz** [comment]
I'll review it tomorrow as well! (now that I've successfully defended, I have a bit more time!)

**Comment by jayaddison** [comment]
Well done completing the defence @picnixz :)  Thank you!

File: tests/test_search.py
Comment (picnixz) [reply tests/test_search.py:186]: So `_title_mapping` are unsorted in the tests, but in production they are sorted? if so, a comment is needed (at least to remember the reason).

**Review by picnixz — APPROVED** [review]
I think my last interrogation was resolved. I'm also happy that we don't have those duplicated entries anymore! I'll wait for Will to also review it and then we can merge it (just ping me when it's done).

File: tests/test_search.py
Comment (jayaddison) [reply tests/test_search.py:186]: Yep, pretty much.  I'll be a bit verbose: 

* In-memory (and therefore in the tests), each value in `_title_mapping` may store elements in arbitrary ordering.
* The tests use Python `set` equality comparison, and that is order-agnostic (as it should be) -- so the tests aren't really checking the order, even though it might appear from the code that they would.
* The output of a Sphinx production build to output a `searchindex.js` sorts these entries.

What you suggest about a comment makes sense...

...however.. perhaps I could attempt a separate PR to refactor the code so that sorting occurs in-place and in-memory, and to adjust the unit tests to assert on that.  Those would provide stronger properties than a comment in the code.

Comment (picnixz) [reply tests/test_search.py:186]: No need for that. It's better that at runtime, the objects are represented in sets since they are more efficient. It's only for serialization that you need to sort to ensure reproducibility.

Comment (jayaddison) [reply tests/test_search.py:186]: Ok, agreed (storing the lists in-place for ordered iteration would probably add various types of overhead).

I've pushed a commit to add some commentary as a docstring on the relevant method and also in the unit tests.

**Review by wlach — APPROVED** [review]
All looks good from my end! @picnixz please merge when ready

**Comment by picnixz** [comment]
I'll do it tomorrow (I was working on something else and missed the notification but I'm going offline now and I prefer not to merge something that I won't be able to see the merge result of).

**Comment by picnixz** [comment]
Thank you Jay!

**Comment by jayaddison** [comment]
Thank you @picnixz!

**Pull request merged**
